### ISSUE TITLE
port mapping improvement?

### ISSUE THREAD

**Issue by c0b** [moby/moby#34019]
current port mapping support this format: `host-ip:hostPort:containerPort/protocol` I am seeing a few things can be improved here:

1. I tried an IPv6 address as the host-ip, seems not recognizable?
2. the hostPort:containerPort support a port range, but when I do a port range, I see it creates a mapping for every individual port pair, this may be inefficient, as far as I know, iptables support a port range in a single rule; the docker-proxy mode may also be improved to support a port range in single process?
3. I don't see this format support to specify a containerIP to connect, you may say the containerIP (`172.17.0.6` in this example) is determined by the runtime, but what I say is sometimes a third-party program may listen on localhost (`127.0.0.1`) only in the container, can this docker-proxy be improved to take a second parameter of which IP in the container to connect to?
   for example, when I run chrome-browser inside the container, and enable its remote-debugging-port=9222, the chrome would listen on 127.0.0.1 only and since Google releases this chrome as a binary, I don't have an option to modify it and don't want to; here need a proxy program to listen on the host and forward packet into container but it create connection by another ip address (127.0.0.1) instead of the continerIP (172.17.0.6); so far I can run `socat` to create the proxy connection, but can this mode be integrated into docker-proxy?

https://docs.docker.com/engine/reference/run/#expose-incoming-ports

```console
#### from 'ps f -e'

 4617 ?        Sl     0:00  \_ /usr/bin/docker-proxy -proto tcp -host-ip 0.0.0.0 -host-port 9230 -container-ip 172.17.0.6 -container-port 19230         
 4625 ?        Sl     0:00  \_ /usr/bin/docker-proxy -proto tcp -host-ip 0.0.0.0 -host-port 9229 -container-ip 172.17.0.6 -container-port 19229
 4634 ?        Sl     0:00  \_ /usr/bin/docker-proxy -proto tcp -host-ip 0.0.0.0 -host-port 9228 -container-ip 172.17.0.6 -container-port 19228
 4643 ?        Sl     0:00  \_ /usr/bin/docker-proxy -proto tcp -host-ip 0.0.0.0 -host-port 9227 -container-ip 172.17.0.6 -container-port 19227
 4651 ?        Sl     0:00  \_ /usr/bin/docker-proxy -proto tcp -host-ip 0.0.0.0 -host-port 9226 -container-ip 172.17.0.6 -container-port 19226
 4659 ?        Sl     0:00  \_ /usr/bin/docker-proxy -proto tcp -host-ip 0.0.0.0 -host-port 9225 -container-ip 172.17.0.6 -container-port 19225
 4668 ?        Sl     0:00  \_ /usr/bin/docker-proxy -proto tcp -host-ip 0.0.0.0 -host-port 9224 -container-ip 172.17.0.6 -container-port 19224
 4676 ?        Sl     0:00  \_ /usr/bin/docker-proxy -proto tcp -host-ip 0.0.0.0 -host-port 9223 -container-ip 172.17.0.6 -container-port 19223
 4685 ?        Sl     0:00  \_ /usr/bin/docker-proxy -proto tcp -host-ip 0.0.0.0 -host-port 9222 -container-ip 172.17.0.6 -container-port 19222
```

**Comment by thaJeztah** [comment]
For the port-ranges, that looks like a duplicate of https://github.com/moby/moby/issues/11185 (and https://github.com/moby/moby/issues/19017). The user land proxy is only necessary to provide "hairpin" routing for distros that don't support this. We try to get rid of the proxy altogether (see https://github.com/moby/moby/issues/14856), but are currently blocked on a kernel bug that is more prevalent if the proxy is disabled https://github.com/moby/moby/issues/5618

(Setting the IPTables rules as ranges is indeed something that should be looked at)

> I tried an IPv6 address as the host-ip, seems not recognizable?

What notation did you use for the IPv6 address?

> sometimes a third-party program may listen on localhost (127.0.0.1) only in the container, can this docker-proxy be improved to take a second parameter of which IP in the container to connect to?

No, I don't think that would be possible; the container has it's own networking namespace, and the proxy is not in that namespace; also it's not the purpose of the user land proxy. For these situations, your best bet may be to either;

- run the container with `--network=host`, in which case the container runs in the host's network stack, i.e. `localhost` in the container is `localhost` of the host
- if you don't want to run `socat` in the same container, you can have a helper container run in the same network namespace as the chrome container (`--network=container:<id of the chrome container>`)

**Comment by c0b** [comment]
> What notation did you use for the IPv6 address?

when I changed to the notation of `-p [::1]:3000:3000/tcp` that works; however docker-compose (version 1.14.0, build c7bdf9e) does not recognize that; could be filed to docker-compose issue instead

```console
ERROR: The Compose file './docker-compose.yml' is invalid because:
services.my-service.ports is invalid: Invalid port "[::1]:3000:3000/tcp", should be [[remote_ip:]remote_port[-remote_port]:]port[/protocol]
```

**Comment by thaJeztah** [comment]
Yes, that looks like a bug in the way docker-compose parses that information; definitely worth opening a bug-report for that

**Comment by thaJeztah** [comment]
Thanks!
